File: RogueTraderSystemGenerator/XenosPrimitive.cs
using System.Runtime.Serialization;

namespace RogueTraderSystemGenerator
{
    [DataContract]
    class XenosPrimitive : XenosBase
    {
        [DataMember]
        public DocContentItem UnusualCommunication { get; set; }
        [DataMember]
        public DocContentItem SocialStructure { get; set; }

        public XenosPrimitive()
        {
            UnusualCommunication = new DocContentItem("No");
            SocialStructure = new DocContentItem("None");
        }

        public void Generate()
        {
            GeneratePrimitiveXenos();
            
            switch(Globals.RollD5())
            {
                case 1:
                    Traits.Deadly++;
                    break;
                case 2:
                    Traits.Mighty++;
                    break;
                case 3:
                    Traits.Resilient++;
                    break;
                case 4:
                    Traits.Stealthy++;
                    break;
                case 5:
                    Traits.Swift++;
                    break;
            }

            switch (Globals.RollD10())
            {
                case 1:
                    Traits.Crawler++;
                    break;
                case 2:
                    Traits.Flyer = 6;
                    break;
                case 3:
                    Traits.Hoverer = 4;
                    break;
                case 4:
                    Traits.MultipleArms++;
                    break;
                case 5:
                    Traits.Quadruped++;
                    break;
                case 6:
                    Traits.Size = XenosSizes.Hulking;
                    break;
                case 7:
                    Traits.Size = XenosSizes.Scrawny;
                    break;
                case 8:
                case 9:
                case 10:
                    // Regular humanoid, average size
                    break;
            }

            if(Globals.RollD100() <= 25)
            {
                switch(Globals.RollD10())
                {
                    case 1:
                        Traits.Armoured++;
                        break;
                    case 2:
                        Traits.Disturbing++;
                        break;
                    case 3:
                        Traits.Deathdweller++;
                        break;
                    case 4:
                        Traits.LethalDefences++;
                        break;
                    case 5:
                        Traits.Disturbing++;
                        break;
                    case 6:
                        Traits.Warped++;
                        break;
                    case 7:
                        Traits.Darkling++;
                        break;
                    case 8:
                        Traits.Unkillable++;
                        break;
                    case 9:
                        Traits.ProjectileAttack++;
                        break;
                    case 10:
                        Traits.Deterrent++;
                        break;
                }
            }

            if(Globals.RollD100() <= 25)
            {
                switch (Globals.RollD5())
                {
                    case 1:
                        UnusualCommunication = new DocContentItem("Intuitive Communicators", 137, "Table 4-11", RuleBook.TheKoronusBestiary);
                        break;
                    case 2:
                        UnusualCommunication = new DocContentItem("Previous Contact", 137, "Table 4-11", RuleBook.TheKoronusBestiary);
                        break;
                    case 3:
                        UnusualCommunication = new DocContentItem("Relic Civilisation", 137, "Table 4-11", RuleBook.TheKoronusBestiary);
                        break;
                    case 4:
                        UnusualCommunication = new DocContentItem("Simplistic", 137, "Table 4-11", RuleBook.TheKoronusBestiary);
                        break;
                    case 5:
                        UnusualCommunication = new DocContentItem("Exotic", 137, "Table 4-11", RuleBook.TheKoronusBestiary);
                        break;
                }
            }

            switch (Globals.RollD10())
            {
                case 1:
                case 2:
                    SocialStructure = new DocContentItem("Agriculturalist", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 3:
                    SocialStructure = new DocContentItem("Hunter", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 4:
                    SocialStructure = new DocContentItem("Feudal", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 5:
                    SocialStructure = new DocContentItem("Raiders", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 6:
                    SocialStructure = new DocContentItem("Nomadic", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 7:
                    SocialStructure = new DocContentItem("Hivemind", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 8:
                    SocialStructure = new DocContentItem("Scavengers", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 9:
                    SocialStructure = new DocContentItem("Xenophobic", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
                case 10:
                    SocialStructure = new DocContentItem("Tradition-bound", 137, "Table 4-12", RuleBook.TheKoronusBestiary);
                    break;
            }

            switch (Traits.Size)
            {
                case XenosSizes.Miniscule:
                    Stats.Strength -= 25;
                    Stats.Toughness -= 25;
                    break;
                case XenosSizes.Puny:
                    Stats.Strength -= 20;
                    Stats.Toughness -= 20;
                    break;
                case XenosSizes.Scrawny:
                    Stats.Strength -= 10;
                    Stats.Toughness -= 10;
                    break;
                case XenosSizes.Average:
                    break;
                case XenosSizes.Hulking:
                    Stats.Strength += 5;
                    Stats.Toughness += 5;
                    Stats.Agility -= 5;
                    break;
                case XenosSizes.Enormous:
                    Stats.Strength += 10;
                    Stats.Toughness += 10;
                    Stats.Agility -= 10;
                    break;
                case XenosSizes.Massive:
                    Stats.Strength += 20;
                    Stats.Toughness += 20;
                    Stats.Agility -= 20;
                    break;
            }

            CalculateEffectsFromTraits();
            CalculateMovement();
        }

        private void GeneratePrimitiveXenos()
        {
            // Generate base stats
            Stats.WeaponSkill = 35;
            Stats.BallisticSkill = 25;
            Stats.Strength = 30;
            Stats.Toughness = 35;
            Stats.Agility = 30;
            Stats.Intelligence = 30;
            Stats.Perception = 35;
            Stats.WillPower = 30;
            Stats.Fellowship = 25;

            BaseMovement = 3;
            Wounds = 10;

            // Generate Skills
            Skills.AwarenessPer++;
            Skills.SurvivalInt = 2;
            Skills.WranglingInt++;

            // Generate Talents

            // Generate Traits

            // Generate Weapons (Note that weapon damage bonus is NOT including strength)
            Weapon weaponSpear = new Weapon("Hunting spear", false, 1, 0, "R", 0, true) {IsMeleeThrown = true, IsNaturalWeapon = false};
            Weapons.Add(weaponSpear);
            Weapon weaponClub = new Weapon("Heavy club", true, 1, 1, "I", 0, true) {IsNaturalWeapon = false};
            Weapons.Add(weaponClub);
        }

        private void CalculateEffectsFromTraits()
        {
            if (Traits.Apex > 0)
            {
                Stats.WeaponSkill += 10;
                Stats.Strength += 10;
                Stats.Toughness += 10;
                Stats.Agility += 10;
                Stats.Perception += 10;
                Talents.SwiftAttack++;
                Traits.ImprovedNaturalWeapons++;

                if (Traits.Apex > 1)
                {
                    Traits.UnnaturalStrength++;
                    Traits.UnnaturalToughness++;
                }
            }

            if (Traits.Amorphous > 0)
            {
                _amorphousMovement = true;
                Stats.Toughness += 10;
                Traits.StrangePhysiology++;
                Traits.UnnaturalSenses += 15;
                if (Traits.Fear >= 2)
                    Traits.Fear++;
                else
                    Traits.Fear = 2;
                if (Globals.RollD100() <= 50)
                    Skills.ClimbSt++;
                if (Globals.RollD100() <= 50)
                    Skills.SwimSt++;
            }

            if (Traits.Amphibious > 0)
            {
                Skills.SwimSt = 3;
            }

            if (Traits.Arboreal > 0)
            {
                Skills.AcrobaticsAg = 3;
                Skills.ClimbSt = 3;
                Skills.DodgeAg = 3;
                Talents.Catfall++;
            }

            if (Traits.Armoured > 0)
            {
                for (int i = 0; i < Traits.Armoured; i++)
                {
                    if (Traits.NaturalArmour <= 0)
                        Traits.NaturalArmour = Globals.RollD5();
                    else
                    {
                        Traits.NaturalArmour += 2;
                        if (Traits.NaturalArmour > 8)
                            Traits.NaturalArmour = 8;
                    }
                }
            }

            if (Traits.Crawler > 0)
            {
                const int baseChance = 25;
                int baseBonus = 5 * (Traits.Crawler - 1);
                if (baseBonus > 20)
                    baseBonus = 20;
                if (Globals.RollD100() <= baseChance + baseBonus)
                    Skills.ClimbSt++;
            }

            if (Traits.Darkling > 0)
            {
                Traits.Blind++;
                if (Globals.RollD100() <= 50)
                    Traits.SonarSense++;
                else
                    Traits.UnnaturalSenses += 30;
                Skills.AwarenessPer++;
                Skills.SilentMoveAg++;
                if (Globals.RollD10() <= 3)
                {
                    Skills.ClimbSt++;
                    Skills.SwimSt++;
                }
                if (Traits.Fear < 1 && Globals.RollD100() <= 10)
                    Traits.Fear = 1;
            }

            if (Traits.Deadly > 0)
            {
                Stats.WeaponSkill += 10;
                Traits.ImprovedNaturalWeapons++;
                if (Traits.Deadly > 1)
                {
                    foreach (Weapon weapon in Weapons)
                    {
                        if(weapon.IsNaturalWeapon)
                            weapon.RazorSharp = true;
                    }
                }
                if (Traits.Deadly > 2)
                {
                    Stats.WeaponSkill += 10 * (Traits.Deadly - 2);
                }
            }

            if (Traits.Deathdweller > 0)
            {
                Talents.ResistanceRadiation++;
                Wounds += 3;

                for (int i = 1; i < Traits.Deathdweller; i++)
                {
                    Stats.Toughness += 5;
                    Wounds += 2;
                }
            }

            if (Traits.Disturbing > 0)
            {
                for (int i = 0; i < Traits.Disturbing; i++)
                {
                    Traits.Fear++;
                }
            }

            if (Traits.FadeKind > 0)
            {
                if (Globals.RollD100() <= 50)
                    Traits.Incorporeal++;
                else
                    Traits.Phase++;
                if (Globals.RollD100() <= 25)
                    Traits.Fear++;
            }

            if (Traits.Flexible > 0)
            {
                foreach (Weapon weapon in Weapons)
                {
                    if (weapon.IsNaturalWeapon)
                        weapon.Flexible = true;
                }
                if (Skills.DodgeAg > 0)
                    Skills.DodgeAg++;
                else
                    Skills.DodgeAg = 2;

                if (Traits.Flexible > 1)
                {
                    foreach (Weapon weapon in Weapons)
                    {
                        if (weapon.IsNaturalWeapon)
                            weapon.Snare = true;
                    }
                }

                for (int i = 2; i < Traits.Flexible; i++)
                {
                    Stats.Agility += 10;
                }
            }

            if (Traits.Gestalt > 0)
            {
                Stats.Toughness += 10;
                Stats.WillPower += 10;
                Stats.Intelligence -= 10;
            }

            if (Traits.Mighty > 0)
            {
                Stats.Strength += 10;
                if (Traits.Mighty > 1)
                    Traits.UnnaturalStrength++;
            }

            if (Traits.Paralytic > 0)
            {
                foreach (Weapon weapon in Weapons)
                {
                    if (weapon.IsNaturalWeapon)
                        weapon.Toxic = true;
                }
            }

            if (Traits.ProjectileAttack > 0)
            {
                if (Stats.BallisticSkill < 30)
                    Stats.BallisticSkill = 30;
                Weapon weapon = new Weapon("Projectile attack", false, 1, 3, "I, R, or E", 0, true) {Range = 15};
                for (int i = 1; i < Traits.ProjectileAttack; i++)
                {
                    weapon.Range += 10;
                    weapon.DamageBonus++;
                    weapon.Penetration++;
                    Stats.BallisticSkill += 10;
                }
                Weapons.Add(weapon);
            }

            if (Traits.Resilient > 0)
            {
                Stats.Toughness += 10;
                if (Traits.Resilient > 1)
                    Traits.UnnaturalToughness++;
            }

            if (Traits.Silicate > 0)
            {
                Stats.Agility -= 10;
                int randomArmour = Globals.RollD5() + 1;
                if (randomArmour > Traits.NaturalArmour)
                    Traits.NaturalArmour = randomArmour;
                Traits.UnnaturalStrength++;
                Traits.UnnaturalToughness++;
            }

            if (Traits.Stealthy > 0)
            {
                Skills.ConcealmentAg = 3;
                Skills.ShadowingAg = 3;
                Skills.SilentMoveAg = 3;
                if (Traits.Stealthy > 1)
                {
                    Skills.ConcealmentAg = 4;
                    Skills.ShadowingAg = 4;
                    Skills.SilentMoveAg = 4;
                }
            }

            if (Traits.SustainedLife > 1)
            {
                if (Traits.FlyerAgilityModified <= 0)
                    Traits.FlyerAgilityModified = 1;
            }

            if (Traits.Swift > 0)
            {
                Stats.Agility += 10;
                for (int i = 1; i < Traits.Swift; i++)
                {
                    Traits.UnnaturalSpeed++;
                }
            }

            if (Traits.ThermalAdaptionCold > 0)
            {
                Stats.Toughness += 5;
                Talents.ResistanceCold++;
            }

            if (Traits.ThermalAdaptionHeat > 0)
            {
                Stats.Toughness += 5;
                Talents.ResistanceHeat++;
            }

            if (Traits.Tunneller > 0)
            {
                Traits.Burrower = GetTotalStrengthBonus();
                for (int i = 1; i < Traits.Tunneller; i++)
                {
                    Traits.Burrower += 2;
                }
            }

            if (Traits.Unkillable > 0)
            {
                Wounds += 5 + (Traits.Unkillable - 1) * 2;
                Traits.Regeneration = Traits.Unkillable;
            }

            if (Traits.UprootedMovement > 0)
            {
                _doesNotMove = false;
            }

            if (Traits.Venomous > 0)
            {
                foreach (Weapon weapon in Weapons)
                {
                    if (weapon.IsNaturalWeapon)
                        weapon.Toxic = true;
                }
            }

            if (Traits.MultipleArms > 0)
            {
                Stats.Toughness += 10;
            }

            // Final calculations
            if (Traits.ImprovedNaturalWeapons > 0)
            {
                foreach (Weapon weapon in Weapons)
                {
                    if (weapon.IsNaturalWeapon)
                    {
                        weapon.Primitive = false;
                        if (Traits.ImprovedNaturalWeapons > 1)
                            weapon.DamageBonus += 2;
                    }
                }
            }

            while (Talents.SwiftAttack > 1)
            {
                Talents.SwiftAttack--;
                if (Talents.LightningAttack > 0)
                    Stats.WeaponSkill += 10;
                else
                    Talents.LightningAttack++;
            }

            if (Traits.FlyerAgilityModified > 0)
            {
                int flyValue = Stats.AgilityBonus * Traits.FlyerAgilityModified;
                if (flyValue > Traits.Flyer)
                    Traits.Flyer = flyValue;
                Traits.FlyerAgilityModified = 0;
            }
        }

    }
}

File: electron-app/js/data/xenosPrimitive.js
// xenosPrimitive.js - Data-only class for Primitive Xenos
(function(){
    const { RuleBooks, DocReference } = window.CommonData;
    class XenosPrimitiveData {
        constructor() {
            this.unusualCommunication = 'No';
            this.socialStructure = 'None';
            this.stats = {
                weaponSkill: 35,
                ballisticSkill: 25,
                strength: 30,
                toughness: 35,
                agility: 30,
                intelligence: 30,
                perception: 35,
                willPower: 30,
                fellowship: 25
            };
            this.skills = [];
            this.talents = [];
            this.traits = [];
            this.weapons = [];
            this.wounds = 10;
            this.movement = '';
            this.armour = '';
            this.referenceData = null; // { basePrimitive, traits: DocReference[] }
        }

        generate() {
            this._generatePrimitiveXenos();

            // Random trait from D5 roll (from WPF source)
            switch (RollD5()) {
                case 1: this.traits.push('Deadly'); break;
                case 2: this.traits.push('Mighty'); break;
                case 3: this.traits.push('Resilient'); break;
                case 4: this.traits.push('Stealthy'); break;
                case 5: this.traits.push('Swift'); break;
            }

            // Physical characteristics from D10 roll
            switch (RollD10()) {
                case 1: this.traits.push('Crawler'); break;
                case 2: this.traits.push('Flyer (6)'); break;
                case 3: this.traits.push('Hoverer (4)'); break;
                case 4: this.traits.push('Multiple Arms'); break;
                case 5: this.traits.push('Quadruped'); break;
                case 6: this.traits.push('Size (Hulking)'); break;
                case 7: this.traits.push('Size (Scrawny)'); break;
                case 8: case 9: case 10: break; // Average humanoid
            }

            // 25% chance of additional trait (C# parity: mapping 9 Projectile Attack, 10 Deterrent)
            if (RollD100() <= 25) {
                switch (RollD10()) {
                    case 1: this.traits.push('Armoured'); break;
                    case 2: this.traits.push('Disturbing'); break;
                    case 3: this.traits.push('Deathdweller'); break;
                    case 4: this.traits.push('Lethal Defences'); break;
                    case 5: this.traits.push('Disturbing'); break; // duplicate intentional per source table
                    case 6: this.traits.push('Warped'); break;
                    case 7: this.traits.push('Darkling'); break;
                    case 8: this.traits.push('Unkillable'); break;
                    case 9: this.traits.push('Projectile Attack'); break;
                    case 10: this.traits.push('Deterrent'); break;
                }
            }

            // Generate communication method (25% chance in C#; table of 5 results)
            if (RollD100() <= 25) {
                switch (RollD5()) {
                    case 1: this.unusualCommunication = 'Intuitive Communicators'; break;
                    case 2: this.unusualCommunication = 'Previous Contact'; break;
                    case 3: this.unusualCommunication = 'Relic Civilisation'; break;
                    case 4: this.unusualCommunication = 'Simplistic'; break;
                    case 5: this.unusualCommunication = 'Exotic'; break;
                }
            }

            // Always assign social structure (d10 table) for parity with C#
            switch (RollD10()) {
                case 1: case 2: this.socialStructure = 'Agriculturalist'; break;
                case 3: this.socialStructure = 'Hunter'; break;
                case 4: this.socialStructure = 'Feudal'; break;
                case 5: this.socialStructure = 'Raiders'; break;
                case 6: this.socialStructure = 'Nomadic'; break;
                case 7: this.socialStructure = 'Hivemind'; break;
                case 8: this.socialStructure = 'Scavengers'; break;
                case 9: this.socialStructure = 'Xenophobic'; break;
                case 10: this.socialStructure = 'Tradition-bound'; break;
            }

            this._calculateMovement();

            // Build reference data (Primitive Xenos uses Stars of Inequity tables)
            this.referenceData = this._buildReferences();

            // Apply mechanical effects (minimal parity subset with C#)
            this._applyMechanicalEffects();

            // Recalculate movement after stat & trait modifications
            this._calculateMovement();
        }

        _buildReferences(){
            const basePrimitive = DocReference('Primitive Xenos', 373, 'Primitive Xenos', RuleBooks.StarsOfInequity);
            // Minimal trait references covering traits used here
            const map = {
                naturalWeapons: { page: 366, book: RuleBooks.CoreRuleBook, ruleName: 'Natural Weapons' },
                primitive:      { page: 126, book: RuleBooks.TheKoronusBestiary, ruleName: 'Primitive' },
                crawler:        { page: 141, book: RuleBooks.TheKoronusBestiary },
                flyer:          { page: 365, book: RuleBooks.CoreRuleBook },
                hoverer:        { page: 364, book: RuleBooks.CoreRuleBook },
                multipleArms:   { page: 366, book: RuleBooks.CoreRuleBook },
                quadruped:      { page: 367, book: RuleBooks.CoreRuleBook },
                sizeHulking:    { page: 367, book: RuleBooks.CoreRuleBook, ruleName: 'Size (Hulking)' },
                sizeScrawny:    { page: 367, book: RuleBooks.CoreRuleBook, ruleName: 'Size (Scrawny)' },
                armoured:       { page: 140, book: RuleBooks.TheKoronusBestiary },
                disturbing:     { page: 141, book: RuleBooks.TheKoronusBestiary },
                deathdweller:   { page: 141, book: RuleBooks.TheKoronusBestiary },
                lethalDefences: { page: 142, book: RuleBooks.TheKoronusBestiary },
                warped:         { page: 144, book: RuleBooks.TheKoronusBestiary },
                darkling:       { page: 141, book: RuleBooks.TheKoronusBestiary },
                unkillable:     { page: 143, book: RuleBooks.TheKoronusBestiary },
                venomous:       { page: 144, book: RuleBooks.TheKoronusBestiary },
                toxic:          { page: 368, book: RuleBooks.CoreRuleBook },
                swift:          { page: 143, book: RuleBooks.TheKoronusBestiary }
            };
            const norm = (t)=>{
                const s=t.toLowerCase();
                if (s.startsWith('natural weapons')) return 'naturalWeapons';
                if (s.startsWith('primitive')) return 'primitive';
                if (s.startsWith('crawler')) return 'crawler';
                if (s.startsWith('flyer (')) return 'flyer';
                if (s.startsWith('hoverer')) return 'hoverer';
                if (s.startsWith('multiple arms')) return 'multipleArms';
                if (s.startsWith('quadruped')) return 'quadruped';
                if (s.startsWith('size (hulking')) return 'sizeHulking';
                if (s.startsWith('size (scrawny')) return 'sizeScrawny';
                if (s.startsWith('armoured')) return 'armoured';
                if (s.startsWith('disturbing')) return 'disturbing';
                if (s.startsWith('deathdweller')) return 'deathdweller';
                if (s.startsWith('lethal defences')) return 'lethalDefences';
                if (s.startsWith('warped')) return 'warped';
                if (s.startsWith('darkling')) return 'darkling';
                if (s.startsWith('unkillable')) return 'unkillable';
                if (s.startsWith('venomous')) return 'venomous';
                if (s.startsWith('toxic')) return 'toxic';
                if (s.startsWith('swift')) return 'swift';
                return '';
            };
            const traitRefs=[];
            for(const tr of this.traits){
                const key=norm(tr);
                if(!key || !map[key]) continue;
                const m=map[key];
                traitRefs.push(DocReference(tr, m.page, m.ruleName||'', m.book));
            }
            return { basePrimitive, traits: traitRefs };
        }

        _generatePrimitiveXenos() {
            // Fixed baseline stats (C# parity)
            this.stats.weaponSkill = 35;
            this.stats.ballisticSkill = 25;
            this.stats.strength = 30;
            this.stats.toughness = 35;
            this.stats.agility = 30;
            this.stats.intelligence = 30;
            this.stats.perception = 35;
            this.stats.willPower = 30;
            this.stats.fellowship = 25;

            // Fixed wounds baseline
            this.wounds = 10;

            // Fixed skills list (Awareness, Survival +10, Wrangling)
            this.skills = ['Awareness (Per)','Survival +10 (Int)','Wrangling (Int)'];

            // Traits
            this.traits = ['Natural Weapons', 'Primitive'];

            // Weapons
            this.weapons = [
                'Hunting spear (Melee/Thrown; 10m; 1d10+SB R; Pen 0; Primitive)',
                'Heavy club (Melee; 1d10+1+SB I; Pen 0; Primitive)'
            ];
            const additionalWeapons = [
                'Stone axe (Melee; 1d10+2+SB R; Pen 0; Primitive)',
                'Bone knife (Melee; 1d5+SB R; Pen 0; Primitive)',
                'Sling (Ranged; 30m; 1d10+SB I; Pen 0; Primitive)',
                'Blow gun (Ranged; 20m; 1d5 R; Pen 0; Primitive, Toxic)',
                'Net (Thrown; 5m; Snare; Primitive)'
            ];
            if (RollD100() <= 40) this.weapons.push(ChooseFrom(additionalWeapons));

            this.armour = 'Hides (Body 2, Arms 1, Legs 1)';
            // Note: movement is recalculated after trait/size adjustments; base is derived from agility in _calculateMovement
        }

        _calculateMovement() {
            const helper = window.XenosBaseData && window.XenosBaseData.computeMovementForProfile;
            if (helper) {
                this.movement = helper({ agility: this.stats.agility, traits: this.traits, earthScorning: false });
            } else {
                // Fallback to previous logic if helper missing
                let ab = Math.floor(this.stats.agility / 10);
                if (this.traits.includes('Crawler') && ab >=2) ab = Math.floor(ab/2 + (ab % 2));
                if (this.traits.includes('Quadruped')) ab *= 2;
                if (this.traits.includes('Size (Hulking)')) ab += 1; else if (this.traits.includes('Size (Scrawny)')) ab -= 1;
                if (this.traits.includes('Swift')) ab += 1;
                const uns = this.traits.find(t=>t.startsWith('Unnatural Speed'));
                if (uns){ const m=uns.match(/x(\d+)/i); if(m){ ab *= parseInt(m[1],10); }}
                if (ab < 1) ab = 1;
                this.movement = `${ab}/${ab*2}/${ab*3}/${ab*6}`;
            }
        }

        // --- Minimal mechanical parity layer ---
        _applyMechanicalEffects(){
            const has = (prefix)=> this.traits.some(t=> t.toLowerCase().startsWith(prefix.toLowerCase()));

            // Size stat adjustments (only Hulking / Scrawny are possible in primitive tables)
            if (has('size (hulking)')){
                this.stats.strength += 5; this.stats.toughness +=5; this.stats.agility -=5;
            } else if (has('size (scrawny)')){
                this.stats.strength -=10; this.stats.toughness -=10;
            }

            // Single-stack trait effects present in primitive tables
            if (has('mighty')) this.stats.strength +=10;
            if (has('resilient')) this.stats.toughness +=10;
            if (has('swift')) this.stats.agility +=10; // movement recalculated later
            if (has('multiple arms')) this.stats.toughness +=10;

            // Deadly: +10 WS and remove Primitive from natural weapon strings (approximation)
            if (has('deadly')){
                this.stats.weaponSkill +=10;
                // In C#, Deadly improves Natural Weapons (removing Primitive & possibly adding Razor Sharp at higher tiers).
                // Primitive Xenos listed weapons are manufactured tools (spear/club), not Natural Weapons, so we deliberately 
                // do NOT strip 'Primitive' here to avoid incorrectly upgrading gear. If later a natural weapon entry is added
                // (e.g., 'Claws (Melee; 1d10+SB R; Pen 0; Primitive)'), we could detect by naming and selectively remove.
            }

            // Armoured: initial natural armour D5
            if (has('armoured')){
                const armourVal = RollD5();
                this.armour = `All ${armourVal}`;
            }

            // Deathdweller & Unkillable wound mods
            if (has('unkillable')) this.wounds +=5;
            if (has('deathdweller')){ this.wounds +=3; this.stats.toughness +=5; }

            // Projectile Attack weapon
            if (has('projectile attack')){
                this.weapons.push('Projectile attack (Ranged; 15m; 1d10+3 I, R, or E; Pen 0; Primitive)');
            }
        }

        getName(){ return 'Primitive Xenos'; }
    }

    window.XenosPrimitiveData = { XenosPrimitiveData };
})();
